docs/lang/articles/misc/profiler.md

```diff
@@ -4,78 +4,125 @@ sidebar_position: 4
+        x[i] = i
+
+for i in range(8):
+    fill()
+ti.print_kernel_profile_info('trace')
+ti.clear_kernel_profile_info() # clear all records
+
+for i in range(100):
+    fill()
+ti.print_kernel_profile_info() # default mode: 'count'
 ```
 
-`ti.print_profile_info()` prints profiling results in a hierarchical format.
+The outputs would be:
+
+```
+=========================================================================
+X64 Profiler(trace)
+=========================================================================
+[      % |     time    ] Kernel name
+[  0.00% |    0.000  ms] jit_evaluator_0_kernel_0_serial
+[ 60.11% |    2.668  ms] fill_c4_0_kernel_1_range_for
+[  6.06% |    0.269  ms] fill_c4_0_kernel_1_range_for
+[  5.73% |    0.254  ms] fill_c4_0_kernel_1_range_for
+[  5.68% |    0.252  ms] fill_c4_0_kernel_1_range_for
+[  5.61% |    0.249  ms] fill_c4_0_kernel_1_range_for
+[  5.63% |    0.250  ms] fill_c4_0_kernel_1_range_for
+[  5.61% |    0.249  ms] fill_c4_0_kernel_1_range_for
+[  5.59% |    0.248  ms] fill_c4_0_kernel_1_range_for
+-------------------------------------------------------------------------
+[100.00%] Total kernel execution time:   0.004 s   number of records:  9
+=========================================================================
+=========================================================================
+X64 Profiler(count)
+=========================================================================
+[      %     total   count |      min       avg       max   ] Kernel name
+[100.00%   0.033 s    100x |    0.244     0.329     2.970 ms] fill_c4_0_kernel_1_range_for
+-------------------------------------------------------------------------
+[100.00%] Total kernel execution time:   0.033 s   number of records:  1
+=========================================================================
+```
 
 :::note
-`ScopedProfiler` is a C++ class in the core of Taichi. It is not exposed
-to Python users.
+Currently the result of `KernelProfiler` could be incorrect on OpenGL backend due to its lack of support for `ti.sync()`.
 :::
+
+### Advanced mode
+
+For the CUDA backend, `KernelProfiler` has an experimental GPU profiling toolkit, Nvidia CUPTI, which provides low and
+deterministic profiling overhead and is able to capture more than 6000 hardware metrics.
+
+Prerequisites to using CUPTI:
+1. Install CUDA Toolkit.
+2. Add environment variable:
+    `export CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda` to your shell configuration files such as `~/.bashrc` and `~/.zshrc`.
+3. Build Taichi from source with CUDA toolkit:
+    `TAICHI_CMAKE_ARGS="-DTI_WITH_CUDA_TOOLKIT:BOOL=ON" python3 setup.py develop --user`.
+4. Resolve privileges issue of Nvidia profiling module (run with `sudo` to get administrative privileges):
+    Add `options nvidia NVreg_RestrictProfilingToAdminUsers=0` to `/etc/modprobe.d/nvidia-kernel-common.conf`.
+    Then `reboot` should resolve the permision issue (probably needs running `update-initramfs -u` before `reboot`).
```

Comment:
Suggest to also point to [this page](https://developer.nvidia.com/nvidia-development-tools-solutions-err_nvgpuctrperm-permission-issue-performance-counters)